packages/ripple-binary-codec/src/enums/README.md

```diff
@@ -53,8 +57,92 @@ See:
 - https://github.com/ripple/rippled/blob/develop/src/ripple/protocol/TER.h
 - https://xrpl.org/transaction-results.html
 
-TODO: Write a script to read rippled's source file and generate the necessary mapping.
+To generate a new definitions file from rippled source code, use this tool: https://github.com/RichardAH/xrpl-codec-gen
 
 ## Transaction Types
 
 See https://github.com/ripple/rippled/blob/develop/src/ripple/protocol/TxFormats.h
+
+# Defining Your Own Definitions
+
+If you're building your own sidechain or writing an amendment for the XRPL, you may need to create new XRPL definitions.
+
+To do that there are a couple things you need to do:
+
+1. Generate your own `definitions.json` file from rippled source code using [this tool](https://github.com/RichardAH/xrpl-codec-gen) (The default `definitions.json` for mainnet can be found [here](https://github.com/XRPLF/xrpl.js/blob/main/packages/ripple-binary-codec/src/enums/definitions.json))
+2. Create new SerializedType classes for any new Types (So that encode/decode behavior is defined)
+
+- For examples of how to implement that you can look at objects in the [`types` folder](../types/), such as `Amount`, `UInt8`, or `STArray`.
+
+3. Import your `definitions.json` file and `coreTypes` from the `types` folder, then use them to construct your own `XrplDefinitions` object.
+4. Pass the `XrplDefinitions` object whenever you `encode` or `decode` a transaction.
+5. If you added any new transaction types, you should create an `interface` for the transaction that extends `BaseTransaction` from the `xrpl` repo to use it with the functions on `Client` (See the below example of adding a new transaction type)
+
+## Example of adding a new Transaction type
+
+```
+// newTypeDefs is where you can import your custom defined definitions.json file
+const newTypeDefs = require('./new-type.json')
+const { XrplDefinitions, coreTypes, Client } = require('xrpl')
+
+const newDefs = new XrplDefinitions(newTxDefinitions, coreTypes)
```

Comment:
Could `coreTypes` be provided by default? IMO 99.9% of people don't need to define new types, so this just adds additional friction.

Reply:
I'd love to, but I for the life of me can't figure out how to do that and resolve the circular dependency problems :/

Importing `coreTypes` to use it as a default param to `XrplDefinitions` creates 3 circular dependencies:
```
1) packages/ripple-binary-codec/dist/enums/index.js > packages/ripple-binary-codec/dist/enums/xrpl-definitions.js > packages/ripple-binary-codec/dist/types/index.js
2) packages/ripple-binary-codec/dist/enums/index.js > packages/ripple-binary-codec/dist/enums/xrpl-definitions.js > packages/ripple-binary-codec/dist/types/index.js > packages/ripple-binary-codec/dist/types/amount.js > packages/ripple-binary-codec/dist/serdes/binary-parser.js
3) packages/ripple-binary-codec/dist/enums/index.js > packages/ripple-binary-codec/dist/enums/xrpl-definitions.js > packages/ripple-binary-codec/dist/types/index.js > packages/ripple-binary-codec/dist/types/st-array.js > packages/ripple-binary-codec/dist/types/st-object.js
```

In order:
1. This one is caused by attempts to work around the other two circular dependencies. (We import `DEFAULT_DEFINITIONS` into the `coreTypes` file in order to ensure we assign it the `coreTypes` since the circular dependencies often result in `coreTypes` being `undefined` when `DEFAULT_DEFINITIONS` is defined.)
2. The `BinaryParser` relies on knowing how to decode all the types from `DEFAULT_DEFINITIONS, and received that as a global. That is then needed for `Amount`, which is a type in `coreTypes`, which then becomes a default parameter for `XrplDefinitions` and thus `DEFAULT_DEFINITIONS`.
3. Same flow as above, where `STObject` is the type in `coreTypes` this time that needs `DEFAULT_DEFINITIONS`. 

I tried for a couple days, and another couple hours today, but couldn't figure out how to untangle the mess :/ 
(There's this suggestion of making a central import / export file that everything uses so you can manually specify the module resolution order, but that also feels like a bad answer: https://medium.com/visual-development/how-to-fix-nasty-circular-dependency-issues-once-and-for-all-in-javascript-typescript-a04c987cf0de 

I can make the tests pass while _including_ the circular dependency, but it's incredibly hacky: https://github.com/XRPLF/xrpl.js/pull/2178

Basically with the circular dependency we can't tell what order the modules will resolve, and we're guaranteed to get an undefined somewhere since the definitions happens after the imports regardless of which path they follow. This means either 
1.`coreTypes` will be undefined on first load of `DEFAULT_DEFINITIONS`, in which case we need to add the coreTypes again with `associateTypes`.
2. `DEFAULT_DEFINITIONS` will be undefined on first load of `coreTypes`, so we can't `associateTypes`. But that means that when we do load `DEFAULT_DEFINITIONS` coreTypes will be there, so it will be defined in the constructor. 

While doing this passes tests, I can't forsee what other bugs/race conditions it will introduce, so I really would not prefer to do that.

Reply:
Update, I found a way to get the desired behavior. Basically by adding a subclass that imports both `coreTypes` and `XrplDefinitions` (renamed to `XrplDefinitionsBase`) that has a constructor with default value. For end users they can use the new `XrplDefinitions` object which subclasses `XrplDefinitionsBase`.

(Separating `coreTypes` from the `XrplDefinitions` object ended up not making sense because the core issue is that `BinarySerializer` needs to know how to go from random binary to objects, which it does through a combination of fields and types - so you need both.)